File: M3Tools/Types/People/Person.vb
Option Strict On
Namespace Types
	Public MustInherit Class Person
		Inherits DbEntry
		Private __email As MimeKit.MailboxAddress

		<ComponentModel.Category("Basics")>
		<Text.Json.Serialization.JsonPropertyName("firstName")>
		Public Property FirstName As String

		<ComponentModel.Category("Basics")>
		<Text.Json.Serialization.JsonPropertyName("lastName")>
		Public Property LastName As String

		<ComponentModel.Category("Contact")>
		<Text.Json.Serialization.JsonPropertyName("email")>
		Public Property Email As String
			Get
				Return If(String.IsNullOrWhiteSpace(__email.Address), Nothing, __email.Address)
			End Get
			Set(value As String)
				__email = New MimeKit.MailboxAddress(Name, If(value, ""))
			End Set
		End Property

		'<ComponentModel.Browsable(False)>
		<Text.Json.Serialization.JsonPropertyName("name")>
		Public Property Name As String
			Get
				Return String.Join(" "c, FirstName, LastName).Trim()
			End Get
			Set(name As String)
				ParseName(name)
			End Set
		End Property

		Protected Friend Sub New(Optional id As Integer = -1, Optional name As String = "John Doe", Optional email As String = "JohnDoe@domain.ext")
			MyBase.New(id)
			ParseName(name)
			Me.Email = If(email, String.Empty).Trim()
		End Sub

		Private Sub ParseName(name As String)
			' TODO: Look into better parsing name data
			Dim parts = name.Trim().Split(" "c)
			FirstName = parts(0)

			Try
				LastName = parts(1)
			Catch
				LastName = Nothing
			End Try
		End Sub

		Overloads Shared Operator =(ls As Person, rs As Person) As Boolean
			Return ls.Id = rs.Id And ls.Name.Equals(rs.Name) And ls.Email.Equals(rs.Email)
		End Operator

		Overloads Shared Operator <>(ls As Person, rs As Person) As Boolean
			Return Not ls = rs
		End Operator
	End Class
End Namespace

File: M3Tools/Types/People/Person.cs
using System.Linq;

namespace SPPBC.M3Tools.Types
{
	/// <summary>
	/// 
	/// </summary>
	public abstract class Person : DbEntry
	{
		private readonly MimeKit.MailboxAddress __email;

		/// <summary>
		/// The person's first name
		/// </summary>
		[System.ComponentModel.Category("Basics")]
		[Newtonsoft.Json.JsonProperty("firstName")]
		public string FirstName { get; set; }

		/// <summary>
		/// The person's last name
		/// </summary>
		[System.ComponentModel.Category("Basics")]
		[Newtonsoft.Json.JsonProperty("lastName")]
		public string LastName { get; set; }

		/// <summary>
		/// The person's email address
		/// </summary>
		[System.ComponentModel.Category("Contact")]
		[Newtonsoft.Json.JsonProperty("email")]
		public string Email
		{
			get;// => string.IsNullOrWhiteSpace(__email.Address) ? null : __email.Address;
			set;// => __email = new MimeKit.MailboxAddress(Name, value);
		}

		/// <summary>
		/// The person's whole name
		/// </summary>
		// [ComponentModel.Browsable(False)]
		[Newtonsoft.Json.JsonProperty("name")]
		public string Name
		{
			get => string.Join(" ", FirstName, LastName).Trim();
			set => ParseName(value);
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="id"></param>
		/// <param name="name"></param>
		/// <param name="email"></param>
		[Newtonsoft.Json.JsonConstructor]
		protected Person(int id = -1, string name = "New Person", string email = null) : base(id)
		{
			ParseName(name);
			Email = (email ?? string.Empty).Trim();
		}

		private void ParseName(string name)
		{
			// TODO: Look into better parsing name data
			string[] parts = (name ?? string.Empty).Trim().Split(' ');
			FirstName = parts[0];

			// MAYBE: Convert to netcore instead of .NET Framework to use this
			//			LastName = string.Join(" ", parts[1..]);
			//LastName = string.Join(" ", parts[1..]);
			LastName = parts.Length > 1 ? string.Join(" ", parts.Skip(1)) : null;
		}
	}
}
